url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/javascheme/SchemeEval.java

```diff
@@ -0,0 +1,804 @@
+    return SchemeObject.OK_SYMBOL;
+  }
+  
+  //While loops
+  public boolean isWhile(SchemeObject exp){
+    return isTaggedList(exp, SchemeObject.WHILE_SYMBOL);
+  }
+  public SchemeObject whilePredicate(SchemeObject exp){
+    return SchemeObject.cadr(exp);
+  }
+  
+  public SchemeObject whileBody(SchemeObject exp){
+    return SchemeObject.cddr(exp);
+  }
+  
+  public SchemeObject doWhile(SchemeObject exp, SchemeObject env){
+    SchemeObject predicate = whilePredicate(exp);
+    SchemeObject body = whileBody(exp);
+    
+    SchemeObject result = SchemeObject.THE_EMPTY_LIST;
+    while(SchemeObject.isTrue(eval(predicate, env))){
+      SchemeObject exps = body;
+      while(!isLastExp(exps)){
+        eval(firstExp(exps), env);
+        exps = exps.getCdr();
+      }
+      result = eval(firstExp(exps), env);
+    }
+    
+    return result;
+  }
+  
+  //Eval
+  public boolean isEval(SchemeObject exp){
+    return isTaggedList(exp, SchemeObject.EVAL_SYMBOL);
+  }
+  
+  public char[] evalString(SchemeObject exp){
+    return SchemeObject.cadr(exp).getString();
+  }
+  
+  public SchemeObject doEval(SchemeObject exp, SchemeObject env){
+    String stringToEval = new String(evalString(exp));
+    SchemeReader reader = SchemeReader.getStringReader(stringToEval);
+    SchemeObject obj = reader.read();
+    SchemeObject retObj = null;
+    while(obj != null){
+      retObj = eval(obj, env);
+      obj = reader.read();
+    }
+    reader.closeStream();
+    if(retObj == null){
+      return SchemeObject.OK_SYMBOL;
+    }else{
+      return retObj;
+    }
+  }
+  
+  //Environment stuff
+  public SchemeObject enclosingEnvironment(SchemeObject env){
+    return SchemeObject.cdr(env);
+  }
+  public SchemeObject firstFrame(SchemeObject env){
+    return SchemeObject.car(env);
+  }
+  public SchemeObject makeFrame(SchemeObject vars, SchemeObject vals){
+    //return SchemeObject.cons(vars, vals);
+    SchemeObject bindings = SchemeObject.THE_EMPTY_LIST;
+    while(!vars.isEmptyList()){
+      bindings = SchemeObject.cons(vars.getCar(), 
+            SchemeObject.cons(vals.getCar(), bindings));
+      vars = vars.getCdr();
+      vals = vals.getCdr();
+    }
+    return SchemeObject.makeHashMap(bindings);
+  }
+  
+  public SchemeObject frameVars(SchemeObject frame){
+    return SchemeObject.car(frame);
+  }
+  public SchemeObject frameVals(SchemeObject frame){
+    return SchemeObject.cdr(frame);
+  }
+  
+  public void addBindingToFrame(SchemeObject var, SchemeObject val, SchemeObject frame){
+    //frame.setCar(SchemeObject.cons(var, frame.getCar()));
+    //frame.setCdr(SchemeObject.cons(val, frame.getCdr()));
+    frame.getHashMap().put(var, val);
+  }
+  
+  public SchemeObject extendEnvironment(SchemeObject vars, SchemeObject vals,
+                                  SchemeObject base_env){
+    return SchemeObject.cons(makeFrame(vars, vals), base_env);
+  }
+  
+  public SchemeObject lookupVariableValue(SchemeObject var, SchemeObject env){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 624-631 into a separate method.